## README

### The Utility of Mixed-Effects Models in the Evaluation of Complex Genomic Traits In Vitro  (Repository)
**Full Article:** [https://dmd.aspetjournals.org/content/early/2023/08/10/dmd.123.001260](https://doi.org/10.1124/dmd.123.001260)

#### ABSTRACT
In pharmacogenomic studies, the use of human liver microsomes as a model system to evaluate the impact of complex genomic traits (i.e., linkage-disequilibrium patterns, coding, and non-coding variation, etc.) on efficiency of drug metabolism is challenging. To accurately predict the true effect size of genomic traits requires large richly sampled datasets representative of the study population. Moreover, the acquisition of this data can be labor-intensive if the study design or bioanalytical methods are not high throughput, and it is potentially unfeasible if the abundance of sample needed for experiments is limited. To overcome these challenges, we developed a novel strategic approach using non-linear mixed effect models (NLME) to determine enzyme kinetic parameters for individual liver specimens using sparse data. This method can facilitate evaluation of the impact of complex genomic traits on the metabolism of xenobiotics in vitro when tissue and other resources are limited. In addition to facilitating the accrual of data, it allows for rigorous testing of covariates as sources of kinetic parameter variability. In this in silico study, we present a practical application of such an approach using previously published in vitro CYP2D6 data and explore the impact of sparse sampling, and experimental error on known kinetic parameter estimates of CYP2D6 mediated formation of 4-hydroxy-atomoxetine in human liver microsomes.



#### SIGNIFICANCE STATMENT
This study presents a novel NLME based framework for evaluating the impact of complex genomic traits on Michaelis-Menten parameters in vitro using sparse data. The utility of this approach extends beyond gene variant associations, including determination of covariate effects on in vitro kinetic parameters, and reduced demand for precious experimental material.



<a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>.

This repository provides the R code, data files used to conduct the in-silico portion of the published study (The Utility of Mixed-Effect Models in the Evaluation of Complex Genomic Traits). Additionally, summary reports and model diagnostics are provided as pdf and integrated HTML files.
